### ISSUE TITLE
Block size not honoured

### ISSUE THREAD

**Issue by kefir-** [sleuthkit/sleuthkit#752]
I recently worked on an SxS memory card, and it had a 2kb sector size. Autopsy 4.3.0 does not handle this, and that leads me to believe that the latest sleuthkit doesn't either, though I haven't tested it.

I tested with an older version of mmls, and although the man page says that sector size is honoured, it doesn't appear to be. Here's a test disk as fdisk sees it:
```$ sudo fdisk -l /dev/sdb
Disk /dev/sdb: 512 MiB, 536870912 bytes, 262144 sectors
Units: sectors of 1 * 2048 = 2048 bytes
Sector size (logical/physical): 2048 bytes / 4096 bytes
I/O size (minimum/optimal): 4096 bytes / 4096 bytes
Disklabel type: dos
Disk identifier: 0x0cdfa81c

Device     Boot Start    End Sectors  Size Id Type
/dev/sdb1         512 262143  261632  511M  c W95 FAT32 (LBA)
```
Here's mmls' view of the same drive:
```
$ sudo mmls /dev/sdb
DOS Partition Table
Offset Sector: 0
Units are in 512-byte sectors

      Slot      Start        End          Length       Description
000:  Meta      0000000000   0000000000   0000000001   Primary Table (#0)
001:  -------   0000000000   0000000511   0000000512   Unallocated
002:  000:000   0000000512   0000262143   0000261632   Win95 FAT32 (0x0c)
003:  -------   0000262144   0001048575   0000786432   Unallocated
```
If I specify the sector size with -b 2048, things look much better. The same happens with an E01 image of the disk above, but the E01 format contains the correct sector size:

```
$ ewfinfo 2k-image.E01 | grep sector
	Bytes per sector:	2048
	Number of sectors:	262144
```
Here's a sample E01 image with 2kb sector size:

[2k-image.zip](https://github.com/sleuthkit/sleuthkit/files/718062/2k-image.zip)

**Comment by bcarrier** [comment]
Thanks kefir.  If the E01 file has the sector size in it, we should be using that. Is that the case? Does it look OK if you give the E01? I can't tell from your message.   A quick glance at the code suggests it should work. 

`vs->block_size = img_info->sector_size;`


If the same file came in as a raw image though, I'm not sure we can detect that it's bad.  I think the only way to do that would be to open the FAT partition (in this case) and see how big it thinks it is and see if that is consistent with a 512-byte sector, 2k, or 4k.  

I believe that we can detect the size with other partition types because they are at a fixed sector offset from the start of the image and we can then know that the sector size is different.  DOS partitions though are in sector 0...

**Comment by kefir-** [comment]
I don't think img_info->sector_size is populated with the information from libewf, because it doesn't work with the latest release of Autopsy. Here's a comparison of EnCase and Autopsy 4.3.0 on the 2k-image I attached in the first post here. As you can see, Autopsy fails and mentions a 512 byte sector offset, but this image is fine if you just parse it with the 2k sector size that is reported by the E01 file.

![encase-vs-autopsy-sectorsize](https://cloud.githubusercontent.com/assets/684832/22925872/a2a34076-f2aa-11e6-9ff4-c51bbc60d574.png)

Regarding heuristics for detecting if a dd image with a DOS/MBR partition table should be parsed as a 4k image (or even a 2k image, though I think that is quite uncommon), I think it'd be useful to simply look at the partition table layout. Since partitions refer to sector numbers rather than byte offsets, you would probably quickly find that the partitions only fill close to one eighth of the disk when working with 512 byte sector size, if the sector size was actually 4096. And the remaining approximate seven eights of the disk would not be allocated to a any partition.

In my example image above, the last partition ends on sector 262143. However, sectors 262144-1048575 (786432 sectors) are unallocated `1048575 / 262143 = 4.0000114`, so in this case it looks like our sector size could be off by a factor of 4 - which is exactly what is happening.

**Comment by kefir-** [comment]
Of course, the math should read `1048576/262144=4.0`, because we're calculating counts and not the last sector. But due to performance and alignment issues when creating partitions, I guess you can't expect the factor to be exactly 4.0 or 8.0 in all cases.

**Comment by kefir-** [comment]
Thanks, I'll try to test this soon!

**Comment by bcarrier** [comment]
Adding some more notes here before i forget:
- PR #951 fixes this issue for E01 files (if the E01 has the correct sector size)
- We still have the issue for raw/dd images, which gets more complicated because we can't detect the non-4096 byte sector until we start to interpret volume or file systems.
- Specific problems:
-- There are various places we can detect the sector size in the volume system.  GPT already does this and @kefir- points out a test for DOS.  But, that code currently never gets run if someone runs a file system command and specifies the offset, such as 'fls -o 63 IMAGE'.  So, we will need to either add more sector detection code into the file system layer or introduce a new set of logic that does sector-size detection into the Image Layer so that it is always run.  
-- The DB code inserts the sector size into the DB before the volume or file system methods are run, so even if the volume or file system methods correctly detect the non-512-byte sectors, the DB will currently have 512.
-- The JNI code does not currently pass in the sector size from the DB when it opens the image.  So, it will try to guess it each time instead of honoring what was stored in the DB.  This is the easiest fix in the list. 

I think it would be safest to add more logic into the 'raw_open' method to detect sector-sizes.  But, it gets messy because we'll need to use volume and file-system specific methods for detecting it.   This means we'll either need to add new "test" methods for the various file systems or dual purpose the open methods.  This will also have a performance impact for each command line tool, so it will be more important to specify the sector size each time to avoid the autodetection code.

**Comment by kefir-** [comment]
Some tools, such as EnCase, require the user to specify details such as sector size when adding a raw disk image. Perhaps autopsy should have some input fields to specify these values when adding raw images, all defaulting to reasonable defaults so that they usually don't need much attention.
